### PR TITLE
[Feature] Support histogram aggregation function

### PR DESCRIPTION
**Pull request by HangyuanLiu**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 19 / 19 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|19|19|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/catalog/FunctionSet.java [hunk 1]
```diff
@@ -370,7 +370,7 @@ public class FunctionSet {
 
     // JSON functions
     public static final Function JSON_QUERY_FUNC = new Function(
-            new FunctionName(JSON_QUERY), new Type[] {Type.JSON, Type.VARCHAR}, Type.JSON, false);
+            new FunctionName(JSON_QUERY), new Type[]{Type.JSON, Type.VARCHAR}, Type.JSON, false);
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/catalog/FunctionSet.java:373]: space

File: be/src/exprs/agg/histogram.h [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            v = down_cast<const ColumnType*>(data_column->data_column().get())->get_data()[row_num];
+        } else {
+            v = down_cast<const ColumnType*>(columns[0])->get_data()[row_num];
+        }
+
+        if (this->data(state).buckets.size() == 0) {
+            Bucket<T> bucket(v, v, 1, 1);
+            this->data(state).buckets.push_back(bucket);
+        } else {
+            Bucket<T>* lastBucket = &this->data(state).buckets.back();
```
Comment (Seaven) [inline be/src/exprs/agg/histogram.h:76]:
```suggestion
            Bucket<T> lastBucket = this->data(state).buckets.back();
```

File: be/src/exprs/agg/histogram.h [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+template <typename T>
+struct Bucket {
+public:
+    Bucket() {}
+    Bucket(T lower, T upper, size_t count, size_t upper_repeats)
+            : lower(lower), upper(upper), count(count), upper_repeats(upper_repeats), count_in_bucket(1) {}
+    T lower;
+    T upper;
+    // Up to this bucket, the total value
+    size_t count;
```
Comment (kangkaisen) [inline be/src/exprs/agg/histogram.h:24]: is int32_t enough?

File: be/src/exprs/agg/histogram.h [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                return;
+            }
+
+            const auto* data_column = down_cast<const NullableColumn*>(columns[0]);
+            v = down_cast<const ColumnType*>(data_column->data_column().get())->get_data()[row_num];
+        } else {
+            v = down_cast<const ColumnType*>(columns[0])->get_data()[row_num];
+        }
+
+        if (this->data(state).buckets.size() == 0) {
```
Comment (kangkaisen) [inline be/src/exprs/agg/histogram.h:72]: is_empty

File: be/src/exprs/agg/histogram.h [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+            Bucket<T> bucket(v, v, 1, 1);
+            this->data(state).buckets.push_back(bucket);
```
Comment (kangkaisen) [inline be/src/exprs/agg/histogram.h:74]: Could do some reserve.

File: be/src/exprs/agg/histogram.h [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+    void convert_to_serialize_format(FunctionContext* ctx, const Columns& src, size_t chunk_size,
+                                     ColumnPtr* dst) const override {
+        //Histogram aggregation function only support one stage Agg
+        DCHECK(false);
+    }
+
+    std::string toBucketJson(std::string lower, std::string upper, size_t count, size_t upper_repeats,
+                             double sample_ratio) const {
+        std::string bucket_json_array = "[\"" + lower + "\",\"" + upper + "\",\"" +
```
Comment (kangkaisen) [inline be/src/exprs/agg/histogram.h:122]: could use fmt:format

**Review by Youngwb — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 19 / 19 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|19|19|100.00%|

File: be/src/exprs/agg/histogram.h [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+                    lastBucket->upper_repeats = 1;
+                }
+            }
+        }
+    }
+
+    void update_batch_single_state(FunctionContext* ctx, AggDataPtr __restrict state, const Column** columns,
+                                   int64_t peer_group_start, int64_t peer_group_end, int64_t frame_start,
+                                   int64_t frame_end) const override {
+        //Histogram aggregation function only support one stage Agg
```
[the remark this answers is not in the document]
Comment (HangyuanLiu) [reply be/src/exprs/agg/histogram.h:99]: When FE generates a query plan, it is guaranteed that only one-stage aggregation can be generated

**Review by Youngwb — DISMISSED** [review]

**Comment by HangyuanLiu** [comment]
run starrocks_fe_unittest

**Comment by HangyuanLiu** [comment]
run starrocks_fe_unittest

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 19 / 19 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|19|19|100.00%|

**Review by Youngwb — APPROVED** [review]

**Pull request merged**